proposals/shared-everything-threads/Overview.md

```diff
@@ -234,33 +370,291 @@ fingerprinting. Note that:
+like a release barrier between when a new allocation is initialized and when its address is made
+available to the program.
+
+[release-acquire]: https://en.cppreference.com/w/cpp/atomic/memory_order#Release-Acquire_ordering
+
+### JS API
+
+#### Conversions
+
+The entire point of the `shared` annotation is to prevent unshareable embedder values from being
+shared across threads. As such, `ToWebAssemblyValue(v, type)` is extended so that if `type` is a
+shared reference, a `TypeError` is almost always thrown with these exceptions:
+
+ - Any value that could be converted to a reference to `i31`, `none`, `nofunc`, or `noextern` can
+   also be converted to a `(shared i31)`, `(shared none)`, `(shared nofunc)`, or `(shared noextern)`
+   reference, respectively.
+ - JS Strings can be converted to shared externref since they are shareable according to the [JS
+   shared structs][proposal-structs] proposal.
+ - Shared JS structs, shared JS arrays, `Atomics.Mutex` and any other shareable type introduced by
+   the JS shared structs proposal can be converted to shared externref.
+
+> Note: It would be great for Wasm-JS interoperability if JS shared structs and arrays could also be
+> converted to Wasm shared structs and arrays and vice versa. We will have to see what is possible
+> there.
+
+Conversely, `ToJSValue(w)` should convert WebAssembly shared references to shareable opaque objects
+on the JS side, analogous to how it converts non-shared references to opaque JS objects, at least
+until we get more ergonomic interop between JS and WasmGC. These opaque JS objects will be shared
+objects that do not require rewrapping, and their identities survive round trips via postMessage
+across thread boundaries.
+
+Waitqueue references are converted to `WebAssembly.WaitQueue` objects, described below.
+
+[proposal-structs]: https://github.com/tc39/proposal-structs
+
+#### Shared annotations
+
+Just like the original threads proposal exposed shared memories in the JS API, this proposal exposes
+shared functions, tables, globals, tags, and exceptions. Unlike shared `WebAssembly.Memory` objects,
+which are actually realm-local wrappers around SharedArrayBuffers (which are in turn realm-local
+wrappers around the underlying shared buffers), the new shared tables, globals, tags, and exceptions
+will be shared JS objects that do not require rewrapping. Such shared objects have fixed layout and
+behave mostly like [sealed] JS objects, and their identities survive round trips via postMessage
+across thread boundaries. Additionally, to facilitate calling methods on these shared objects, they
+have [realm-local prototypes][tls-prototypes].
+
+The shared versions of the JS API types (besides memories) will be represented with new
+constructors:
+
+ - `WebAssembly.SharedFunction`
+ - `WebAssembly.SharedTable`
+ - `WebAssembly.SharedGlobal`
+ - `WebAssembly.SharedTag`
+ - `WebAssembly.SharedException`
```

Comment:
I think it can go either way.

The reasoning for the bifurcation is that JS reflections of shared objects are fundamentally different from unshared objects on account of being accessible from multiple agents. This is a deeper sense of shared than what `SharedArrayBuffer` provides, which is a non-shared wrapper object that points to some shared backing store that gets re-wrapped every time it gets sent to a new agent.

Since there's no rewrapping here, and the shared JS reflections aren't extensible, the thinking was that having a different constructor helps drive home the point that these objects are in fact very different.

In practice it probably comes down to does the user want to write code like 

```javascript
if (table instanceof WebAssembly.Table) {
  if (table.shared) {
    handleShared(table);
  } else {
    handleNotShared(table);
  }
}
```

or 

```javascript
if (table instanceof WebAssembly.SharedTable) {
  handleShared(table);
} else if (table instanceof WebAssembly.Table) {
  handleNotShared(table);
}
```

(Or possibly neither, because you've organized your code to handle the shared/unshared split at a much higher level.)

Reply:
Yes, I appreciate that their representation is rather different, but JS already has many different function representations, doesn't it?

Your example only has a benign difference. Much more interesting is the other direction, where you don't want
```
  for (let imp of WebAssembly.Module.imports(module)) {
    switch (imp.kind) {
      case "table":
        value = new WebAssembly.Table(imp.type);
        break;
      case "memory":
        value = new WebAssembly.Memory(imp.type);
        break;
      case "global":
        value = new WebAssembly.Global(imp.type, undefined);
        break;
      case "function":
        value = new WebAssemlby.Function(imp.type, () => { throw "unimplemented" });
        break;
    }
    ...
  }
```
to devolve into
```
  for (let imp of WebAssembly.Module.imports(module)) {
    switch (imp.kind) {
      case "table":
        if (imp.type.shared)
          value = new WebAssembly.SharedTable(imp.type);
        else
          value = new WebAssembly.Table(imp.type);
        break;
      case "memory":
        if (imp.type.shared)
          value = new WebAssembly.SharedMemory(imp.type);
        else
          value = new WebAssembly.Memory(imp.type);
        break;
      case "global":
        if (imp.type.shared)
          value = new WebAssembly.SharedGlobal(imp.type, undefined);
        else
          value = new WebAssembly.Global(imp.type, undefined);
        break;
      case "function":
        if (imp.type.shared)
          value = new WebAssemlby.SharedFunction(imp.type, () => { throw "unimplemented" });
        else
          value = new WebAssemlby.Function(imp.type, () => { throw "unimplemented" });
        break;
    }
    ...
  }
```
This would not just be tedious but an unnecessary breaking change to code using the API generically.

Splitting up at the JS constructor level would be very much at odds with Wasm's underlying type structure, and that would show up in places like this.

Reply:
> JS already has many different function representations

Yes, though it's a value judgment on how different the different functions are vs how different shared vs unshared objects are.

I don't feel particularly strong here. I'd be happy for this to be decided by practitioners in the Wasm/JS API space (tools?).

Reply:
I don't feel too strongly either. I was thinking that it would be good to make the underlying differences between the shared and unshared objects more explicit by giving them different constructors, but @rossberg, I suspect you're right that the usability story makes more sense without that bifurcation.